### PR TITLE
*: add config manager

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve? 
Part of #1977.

### What is changed and how it works?
This PR adds a config manager to manage all components' config.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: server/config_manager/config_manager.go [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+		if cfg, ok := componentsCfg[componentID]; ok {
+			config, err = c.getComponentCfg(component, componentID)
+			if err != nil {
+				return version, "", &configpb.Status{
+					Code:    configpb.StatusCode_UNKNOWN,
+					Message: ErrEncode(err),
+				}
+			}
+			if reflect.DeepEqual(cfg.getVersion(), version) {
+				status = &configpb.Status{Code: configpb.StatusCode_OK}
```
Comment (NingLin-P) [inline server/config_manager/config_manager.go]: TiKV expect `StatusCode::NotChange` in this situation.

File: server/config_manager/config_manager.go [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+	latestVersion := c.getLatestVersion(component, componentID)
+	initVersion := &configpb.Version{Local: 0, Global: 0}
+	if componentsCfg, ok := c.LocalCfgs[component]; ok {
+		if _, ok := componentsCfg[componentID]; ok {
+			// restart a component
+			if reflect.DeepEqual(initVersion, version) {
+				status = &configpb.Status{Code: configpb.StatusCode_OK}
+			} else {
+				status = &configpb.Status{Code: configpb.StatusCode_WRONG_VERSION}
+			}
```
Comment (NingLin-P) [inline server/config_manager/config_manager.go]: For restart component with default `Version` (i.e both local and global set to 0), we should expect that this component did not persist the `version` and `config` so we should return the latest version and config with `StatusCode_OK`.

File: server/config_manager/config_manager.go [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+				status = &configpb.Status{Code: configpb.StatusCode_UNKNOWN, Message: ErrDecode(err)}
+			} else {
+				componentsCfg[componentID] = lc
+				status = &configpb.Status{Code: configpb.StatusCode_OK}
+			}
+		}
+	} else {
+		c.LocalCfgs[component] = make(map[string]*LocalConfig)
+		// start the first component
+		lc, err := NewLocalConfig(cfg, initVersion)
```
Comment (NingLin-P) [inline server/config_manager/config_manager.go]: Maybe we should also create a new `GlobalConfig` here.

File: server/config_manager/config_manager.go [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+// ApplyGlobalConifg applies the global change to each local component.
+func (c *ConfigManager) ApplyGlobalConifg(globalCfg *GlobalConfig, component string, newGlobalVersion uint64, entries []*configpb.ConfigEntry) error {
+	// get the global config
+	updateEntries := make(map[string]*EntryValue)
+	for _, entry := range entries {
+		globalCfg.updateEntry(entry, &configpb.Version{Global: newGlobalVersion, Local: 0})
+		globalUpdateEntries := c.GlobalCfgs[component].getUpdateEntries()
+		for k, v := range globalUpdateEntries {
+			updateEntries[k] = v
+		}
```
Comment (NingLin-P) [inline server/config_manager/config_manager.go]: Should these code move out of the for loop?

File: server/core/storage.go [hunk 5]
```diff
@@ -333,6 +336,31 @@ func (s *Storage) LoadAllScheduleConfig() ([]string, []string, error) {
 	return keys, values, err
 }
 
+// SaveComponentsConfig stores marshalable cfg to the componentsConfigPath.
+func (s *Storage) SaveComponentsConfig(cfg interface{}) error {
+	value := new(bytes.Buffer)
```
Comment (disksing) [inline server/core/storage.go]: Prefer `var value bytes.Buffer`.

File: server/config_manager/config_manager.go [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+		return fmt.Sprintf("decode error: %v", e)
+	}
+	errNotSupported = "not supported"
+)
+
+// ConfigManager is used to manage all components' config.
+type ConfigManager struct {
+	sync.RWMutex
+	GlobalCfgs map[string]*GlobalConfig
+	LocalCfgs  map[string]map[string]*LocalConfig
```
Comment (disksing) [inline server/config_manager/config_manager.go]: add comment: "component -> componentID -> LocalConfig"

File: server/config_manager/config_manager.go [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+// Reload reloads the configuration from the storage.
+func (c *ConfigManager) Reload(storage *core.Storage) error {
+	_, err := storage.LoadComponentsConfig(c)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *ConfigManager) getComponent(ComponentID string) string {
```
Comment (disksing) [inline server/config_manager/config_manager.go]: ComponentID -> componentID or id

File: server/config_manager/config_manager.go [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+	for component := range c.LocalCfgs {
+		for componentID := range c.LocalCfgs[component] {
+			if componentID == ComponentID {
+				return component
+			}
+		}
+	}
```
Comment (disksing) [inline server/config_manager/config_manager.go]:
```
for component, cfgs := range c.LocalCfgs {
    if _, ok := cfgs[id]; ok {
        return component
    }
}
```

File: server/config_manager/config_manager.go [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+
+// Get returns config and the latest version.
+func (c *ConfigManager) Get(version *configpb.Version, component, componentID string) (*configpb.Version, string, *configpb.Status) {
+	c.RLock()
+	defer c.RUnlock()
+	var config string
+	var err error
+	var status *configpb.Status
+	if componentsCfg, ok := c.LocalCfgs[component]; ok {
+		if cfg, ok := componentsCfg[componentID]; ok {
```
Comment (disksing) [inline server/config_manager/config_manager.go]: too many indent... I prefer
```
xx, ok := xxx
if !ok {
 return err
}
yy, ok := yyy
if !ok {
 return err
}
...
```

File: server/config_manager/config_manager.go [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+			if reflect.DeepEqual(initVersion, version) {
+				status = &configpb.Status{Code: configpb.StatusCode_OK}
+			} else {
+				status = &configpb.Status{Code: configpb.StatusCode_WRONG_VERSION}
+			}
+		} else {
+			// add a new component
+			lc, err := NewLocalConfig(cfg, initVersion)
+			if err != nil {
+				status = &configpb.Status{Code: configpb.StatusCode_UNKNOWN, Message: ErrDecode(err)}
```
Comment (disksing) [inline server/config_manager/config_manager.go]: not return?

Comment (rleungx) [reply server/config_manager/config_manager.go]: returned in the end of this function

**Comment by NingLin-P** [comment]
LGTM

**Comment by rleungx** [comment]
PTAL @disksing

**Pull request merged**